File: sdk-go/tracing.go
package sdk

import (
	"github.com/nats-io/nats.go"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/propagation"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
	"context"
)

const tracerName = "github.com/vanducng/mio/sdk-go"

// natsHeaderCarrier adapts nats.Header to the OTel TextMapCarrier interface.
// W3C traceparent is ASCII-safe and passes through NATS headers unchanged.
type natsHeaderCarrier struct {
	h nats.Header
}

func (c natsHeaderCarrier) Get(key string) string {
	return c.h.Get(key)
}

func (c natsHeaderCarrier) Set(key, val string) {
	c.h.Set(key, val)
}

func (c natsHeaderCarrier) Keys() []string {
	keys := make([]string, 0, len(c.h))
	for k := range c.h {
		keys = append(keys, k)
	}
	return keys
}

// injectTrace starts a PRODUCER span and injects W3C traceparent into msg headers.
// The returned context carries the active span for the caller to end.
func injectTrace(ctx context.Context, tp trace.TracerProvider, subject, msgID string, header nats.Header) (context.Context, trace.Span) {
	tracer := tp.Tracer(tracerName)
	ctx, span := tracer.Start(ctx, "mio.publish",
		trace.WithSpanKind(trace.SpanKindProducer),
		trace.WithAttributes(
			semconv.MessagingSystemKey.String("nats"),
			attribute.String("messaging.destination", subject),
			attribute.String("messaging.message_id", msgID),
		),
	)
	otel.GetTextMapPropagator().Inject(ctx, natsHeaderCarrier{h: header})
	return ctx, span
}

// extractTrace extracts W3C traceparent from msg headers and starts a CONSUMER span.
// The returned context carries the active span for the caller to end.
func extractTrace(ctx context.Context, tp trace.TracerProvider, subject string, header nats.Header) (context.Context, trace.Span) {
	prop := otel.GetTextMapPropagator()
	parentCtx := prop.Extract(ctx, natsHeaderCarrier{h: header})

	tracer := tp.Tracer(tracerName)
	ctx, span := tracer.Start(parentCtx, "mio.consume",
		trace.WithSpanKind(trace.SpanKindConsumer),
		trace.WithAttributes(
			semconv.MessagingSystemKey.String("nats"),
			attribute.String("messaging.destination", subject),
		),
	)
	return ctx, span
}

// endSpanOnError records err on span and sets status; always ends the span.
func endSpanOnError(span trace.Span, err error) {
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
	}
	span.End()
}

// noopTracerProvider returns the global tracer provider, falling back to noop
// if none has been registered. Used when the caller omits WithTracerProvider.
func noopTracerProvider() trace.TracerProvider {
	return otel.GetTracerProvider()
}

// propagatorFromContext uses the globally registered TextMapPropagator.
// Callers that need W3C tracecontext must register it at startup:
//
//	otel.SetTextMapPropagator(propagation.TraceContext{})
func propagatorFromContext() propagation.TextMapPropagator {
	return otel.GetTextMapPropagator()
}

File: sdk-py/mio/tracing.py
"""W3C traceparent propagation over NATS headers for MIO Python SDK.

Uses opentelemetry.propagators.tracecontext.TraceContextTextMapPropagator.
Carrier adapter wraps a dict[str, str] (NATS message headers).

Callers that need W3C tracecontext must register the propagator at startup:
    from opentelemetry.propagators.tracecontext import TraceContextTextMapPropagator
    from opentelemetry import propagate
    propagate.set_global_textmap(TraceContextTextMapPropagator())
"""

from __future__ import annotations

from opentelemetry import trace, propagate, context
from opentelemetry.trace import SpanKind, Span, TracerProvider
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

_TRACER_NAME = "github.com/vanducng/mio/sdk-py"

# Singleton propagator — uses global propagator by default.
_PROPAGATOR = TraceContextTextMapPropagator()


class _DictCarrier:
    """OTel TextMapCarrier adapter over a plain dict (NATS headers)."""

    def __init__(self, headers: dict[str, str]) -> None:
        self._headers = headers

    def get(self, key: str, default: str | None = None) -> str | None:
        # OTel >= 1.30 DefaultGetter calls carrier.get(key, default).
        # Accept the optional default to stay compatible with both old and new OTel.
        return self._headers.get(key, default)

    def set(self, key: str, value: str) -> None:
        self._headers[key] = value

    def keys(self) -> list[str]:
        return list(self._headers.keys())


def inject_trace(
    headers: dict[str, str],
    subject: str,
    msg_id: str,
    tracer_provider: TracerProvider | None = None,
) -> tuple[context.Context, Span]:
    """Start a PRODUCER span and inject W3C traceparent into headers.

    Returns (ctx, span). Caller must end the span when publish completes.
    """
    tp = tracer_provider or trace.get_tracer_provider()
    tracer = tp.get_tracer(_TRACER_NAME)

    ctx = context.get_current()
    span = tracer.start_span(
        "mio.publish",
        kind=SpanKind.PRODUCER,
        context=ctx,
        attributes={
            "messaging.system": "nats",
            "messaging.destination": subject,
            "messaging.message_id": msg_id,
        },
    )
    # OTel 1.x API: set_span_in_context replaces the deprecated use_span().
    ctx_with_span = trace.set_span_in_context(span, ctx)

    carrier = _DictCarrier(headers)
    _PROPAGATOR.inject(carrier, context=ctx_with_span)

    return ctx_with_span, span


def extract_trace(
    headers: dict[str, str],
    subject: str,
    tracer_provider: TracerProvider | None = None,
) -> tuple[context.Context, Span]:
    """Extract W3C traceparent from headers and start a CONSUMER span.

    Returns (ctx, span). Caller must end the span when message processing completes.
    """
    carrier = _DictCarrier(headers)
    parent_ctx = _PROPAGATOR.extract(carrier)

    tp = tracer_provider or trace.get_tracer_provider()
    tracer = tp.get_tracer(_TRACER_NAME)

    span = tracer.start_span(
        "mio.consume",
        context=parent_ctx,
        kind=SpanKind.CONSUMER,
        attributes={
            "messaging.system": "nats",
            "messaging.destination": subject,
        },
    )
    ctx = trace.set_span_in_context(span, parent_ctx)
    return ctx, span
